client/my-sites/site-settings/form-general.jsx

```diff
@@ -113,57 +112,131 @@ const FormGeneral = React.createClass( {
 			amp_is_supported: false,
 			amp_is_enabled: false,
 		} );
-	},
+		this.props.clearDirtyFields();
+		this.props.updateFields( this.getFormSettings( this.props.settings ) );
+	}
 
-	onTimezoneSelect( timezone ) {
-		this.setState( { timezone_string: timezone } );
-	},
+	componentWillReceiveProps( nextProps ) {
+		this._showWarning( nextProps.site );
+
+		if ( nextProps.settings !== this.props.settings ) {
+			let newState = this.getFormSettings( nextProps.settings );
+			//If we have any fields that the user has updated,
+			//do not wipe out those fields from the poll update.
+			newState = omit( newState, nextProps.dirtyFields );
+			nextProps.updateFields( newState );
+		}
+	}
 
 	onRecordEvent( eventAction ) {
 		return this.recordEvent.bind( this, eventAction );
-	},
+	}
 
 	onRecordEventOnce( key, eventAction ) {
 		return this.recordEventOnce.bind( this, key, eventAction );
-	},
+	}
+
+	recordEvent( eventAction ) {
+		debug( 'record event: %o', eventAction );
+		analytics.ga.recordEvent( 'Site Settings', eventAction );
+	}
+
+	/**
+	 * Record an analytics event only once per mounted component instance
+	 * @param  {string} key         - unique key to namespace the event
+	 * @param  {string} eventAction - the description of the action to appear in analytics
+	 */
+	recordEventOnce( key, eventAction ) {
+		debug( 'record event once: %o - %o', key, eventAction );
+		if ( this.state[ 'recordEventOnce-' + key ] ) {
+			return;
+		}
+		this.recordEvent( eventAction );
+		this.setState( { [ 'recordEventOnce-' + key ]: true } );
+	}
+
+	handleRadio = event => {
+		const currentTargetName = event.currentTarget.name,
+			currentTargetValue = event.currentTarget.value;
+
+		this.props.updateFields( { [ currentTargetName ]: currentTargetValue } );
+	};
+
+	handleSubmitForm = event => {
+		if ( ! event.isDefaultPrevented() && event.nativeEvent ) {
+			event.preventDefault();
+		}
+
+		this.submitForm();
+		this.recordEvent( 'Clicked Save Settings Button' );
+	};
+
+	submitForm() {
+		const { fields, site, clearDirtyFields } = this.props;
+		this.props.removeNotice( 'site-settings-save' );
+		this.props.saveSiteSettings( site.ID, fields )
+			.then( () => {
```

Comment:
We should never use the promise returned by an action creator thunk, outside of tests for said action creators.

This breaks [Redux unidirectional data flow](http://redux.js.org/docs/basics/DataFlow.html).

Instead, you could have a separate reducer key to track progress and check for completion in the component's `componentWillReceiveProps` from that reducer state.